zcash_client_sqlite/src/wallet/commitment_tree.rs

```diff
@@ -746,15 +757,98 @@ pub(crate) fn truncate_checkpoints(
     Ok(())
 }
 
+pub(crate) fn put_shard_roots<
+    H: Hashable + HashSer + Clone + Eq,
+    const DEPTH: u8,
+    const SHARD_HEIGHT: u8,
+>(
+    conn: &rusqlite::Transaction<'_>,
+    table_prefix: &'static str,
+    start_index: u64,
+    roots: &[CommitmentTreeRoot<H>],
+) -> Result<(), ShardTreeError<Error>> {
+    if roots.is_empty() {
+        // nothing to do
+        return Ok(());
+    }
+
+    // We treat the cap as a DEPTH-SHARD_HEIGHT tree so that we can make a batch insertion of
+    // root data using `Position::from(start_index)` as the starting position and treating the
+    // roots as level-0 leaves.
+    let cap = LocatedTree::from_parts(
+        Address::from_parts((DEPTH - SHARD_HEIGHT).into(), 0),
+        get_cap(conn, table_prefix).map_err(ShardTreeError::Storage)?,
+    );
+
+    let cap_result = cap
+        .batch_insert(
+            Position::from(start_index),
+            roots.iter().map(|r| {
+                (
+                    r.root_hash().clone(),
+                    Retention::Checkpoint {
+                        id: (),
+                        is_marked: false,
+                    },
+                )
+            }),
+        )
+        .map_err(ShardTreeError::Insert)?
+        .expect("slice of inserted roots was verified to be nonempty");
+
+    put_cap(conn, table_prefix, cap_result.subtree.take_root()).map_err(ShardTreeError::Storage)?;
+
+    for (root, i) in roots.iter().zip(0u64..) {
+        // To avoid deserializing the subtree just to annotate its root, we simply cache the downloaded
+        // root alongside of the persisted subtree and reannotate the root, handling conflicts, at the
+        // time that we deserialize the tree.
+        let mut stmt = conn
+            .prepare_cached(&format!(
+            "INSERT INTO {}_tree_shards (shard_index, subtree_end_height, root_hash, shard_data)
+            VALUES (:shard_index, :subtree_end_height, :root_hash, :shard_data)
+            ON CONFLICT (shard_index) DO UPDATE
+            SET subtree_end_height = :subtree_end_height, root_hash = :root_hash",
+            table_prefix
+        ))
+            .map_err(|e| ShardTreeError::Storage(Either::Right(e)))?;
+
+        // The `shard_data` value will only be used in the case that no tree already exists.
+        let mut shard_data: Vec<u8> = vec![];
+        let tree = PrunableTree::empty().reannotate_root(Some(Rc::new(root.root_hash().clone())));
+        write_shard(&mut shard_data, &tree)
+            .map_err(|e| ShardTreeError::Storage(Either::Left(e)))?;
```

Comment:
It is not at all clear to me why this is useful. AFAICT `reannotate_root` replaces the annotation on the root node of the tree, but only if the root node is a `Node::Parent`; this is an empty tree with a placeholder root node `Node::Nil`, so the `reannotate_root` does nothing.

Reply:
Ah, good point, this should be a `PrunableTree::leaf()` call.